executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,673 @@
+		for i := 0; i < task.outerResult.NumRows(); i++ {
+			rows = append(rows, task.outerResult.GetRow(i))
+		}
+		sort.Slice(rows, func(i, j int) bool {
+			for id, joinKey := range omw.joinKeys {
+				cmp, _, err := omw.compareFuncs[id](omw.ctx, joinKey, joinKey, rows[i], rows[j])
+				terror.Log(err)
+				if cmp != 0 || omw.nextColCompareFilters == nil {
+					return cmp < 0
+				}
+				return omw.nextColCompareFilters.CompareRow(rows[i], rows[j]) < 0
+			}
+			return false
+		})
+		for i := 0; i < len(rows); i++ {
+			c.AppendRow(rows[i])
+		}
+		task.outerResult = c
+	}
+
+	if omw.filter != nil {
+		var err error
+		task.outerMatch, err = expression.VectorizedFilter(omw.ctx, omw.filter, chunk.NewIterator4Chunk(task.outerResult), task.outerMatch)
+		if err != nil {
+			return task, err
+		}
+		task.memTracker.Consume(int64(cap(task.outerMatch)))
+	}
+	task.outerIter = chunk.NewIterator4Chunk(task.outerResult)
+	return task, nil
+}
+
+func (omw *outerMergeWorker) increaseBatchSize() {
+	if omw.batchSize < omw.maxBatchSize {
+		omw.batchSize *= 2
+	}
+	if omw.batchSize > omw.maxBatchSize {
+		omw.batchSize = omw.maxBatchSize
+	}
+}
+
+func (imw *innerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup, cancelFunc context.CancelFunc) {
+	var task *lookUpMergeJoinTask
+	defer func() {
+		wg.Done()
+		if r := recover(); r != nil {
+			cancelFunc()
+		}
+	}()
+
+	for ok := true; ok; {
+		select {
+		case task, ok = <-imw.taskCh:
+			if !ok {
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+
+		err := imw.handleTask(ctx, task)
+		task.doneErr <- err
+		close(task.doneErr)
+	}
+}
+
+func (imw *innerMergeWorker) handleTask(ctx context.Context, task *lookUpMergeJoinTask) error {
+	dLookUpKeys, err := imw.constructDatumLookupKeys(task)
+	if err != nil {
+		return err
+	}
+	dLookUpKeys = imw.dedupDatumLookUpKeys(dLookUpKeys)
+	imw.innerExec, err = imw.readerBuilder.buildExecutorForIndexJoin(ctx, dLookUpKeys, imw.indexRanges, imw.keyOff2IdxOff, imw.nextColCompareFilters)
+	if err != nil {
+		return err
+	}
+	defer terror.Call(imw.innerExec.Close)
+	_, err = imw.fetchNextInnerResult(ctx, task)
+	if err != nil {
+		return err
+	}
+	err = imw.doMergeJoin(ctx, task)
+	return err
+}
+
+func (imw *innerMergeWorker) fetchNewChunkWhenFull(ctx context.Context, task *lookUpMergeJoinTask, chk **chunk.Chunk) bool {
+	if !(*chk).IsFull() {
+		return true
+	}
+	select {
+	case task.results <- &indexMergeJoinResult{*chk, imw.joinChkResourceCh}:
+	case <-ctx.Done():
+		return false
+	}
+	var ok bool
+	*chk, ok = <-imw.joinChkResourceCh
+	if !ok {
+		return false
+	}
+	(*chk).Reset()
+	return true
+}
+
+func (imw *innerMergeWorker) doMergeJoin(ctx context.Context, task *lookUpMergeJoinTask) (err error) {
+	chk := chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+	defer func() {
+		if chk.NumRows() > 0 {
+			select {
+			case task.results <- &indexMergeJoinResult{chk, imw.joinChkResourceCh}:
+			case <-ctx.Done():
+				return
+			}
+		}
```

Comment:
If the chunk is empty, should we still need to put chunk back to `joinChkResourceCh`?

Reply:
Why? I don't know what's your mean.

Reply:
done.